### ISSUE TITLE
[Dev Bug] Referencing a variable that had a semantic error in its definition causes an internal error

### ISSUE THREAD

**Issue by Luna-Klatzer** [Kipper-Lang/Kipper#405]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### I am following the documentation's guide

- [X] I have read the documentation

### This issue exists in the latest version

- [X] I am using the latest version

### Current Behavior

Currently with the latest dev version `0.10.0-alpha.7` there is a bug in the compiler that causes an internal error if a variable is referenced, where the initial definition had an invalid assignment. For example:

```
var x: str = "x"["y"];
x;
```

### Expected Behavior

- The compiler should properly handle this edge case and not throw an internal error when the variable is referenced. 
- The semantic data should be populated as much as possible so that any references don't fail as well and cause confusing errors.

### Steps To Reproduce

1. Create a definition, where the assignment has an error:
    ```
    var x: str = "x"["x"];
    ```
2. Reference the variable using its identifier, like for example:
    ```
    x;
    ```

### Environment

- Kipper: 0.10.0-alpha.7
- Environment: Both Node.js v18 and Firefox 111.0a1
- Operating System: any

**Comment by Luna-Klatzer** [comment]
This likely requires the compiler also to implement a pre-semantic analysis step where all identifiers in the program are registered, so references are determined *before* anything else. Another method would be identifier resolution in a post-semantic analysis step, where all references are resolved in a step between the primary semantic analysis and the type checking.

**Comment by Luna-Klatzer** [comment]
To simplify the error handling, likely a small rewrite of the error recovery system is required. That means instead of throwing errors and aborting a function with an internal error, each semantic analysis function should handle its edge-cases of their children being valid/or not. So for example that a function can check if the children's semantic data is populated and ignore certain checks if they aren't.

This would simplify handling edge cases, but require more checking in each AST node since a general algorithm like the one implemented at the moment won't work anymore and therefore can't handle the errors on a standardised level.

**Comment by Luna-Klatzer** [comment]
Fixed by #408 - Implementation is now a mix of the existing algorithm and the new handling system for declarations/statements. View implementation details here: a49dfe74ff79b8eb94938fc6ca1312670a00e8d7

**Comment by Luna-Klatzer** [comment]
https://github.com/Luna-Klatzer/Kipper/pull/408#issuecomment-1436056334
